R/pipe_consistency_linter.R

```diff
@@ -0,0 +1,79 @@
+#' Pipe consistency linter
+#'
+#' Check that pipe operators are consistently used throughout a file. Optionally
+#' specify which pipe operator to use.
+#'
+#' @param pipe Which pipe operator to use (either `"%>%"` or `"|>"`). By default
+#' (`"auto"``), the linter has no preference but will check that the same pipe
+#' operator is used throughout the file.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "1:3 |> mean() %>% as.character()",
+#'   linters = pipe_consistency_linter()
+#' )
+#'
+#' lint(
+#'   text = "1:3 %>% mean() %>% as.character()",
+#'   linters = pipe_consistency_linter("|>")
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "1:3 %>% mean() %>% as.character()",
+#'   linters = pipe_consistency_linter()
+#' )
+#'
+#' lint(
+#'  text = "1:3 |> mean() |> as.character()",
+#'  linters = pipe_consistency_linter()
+#' )
+#' @evalRd rd_tags("pipe_consistency_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+pipe_consistency_linter <- function(pipe = c("auto", "%>%", "|>")) {
+  pipe <- match.arg(pipe)
+
+  xpath_magrittr <- "//SPECIAL[text() = '%>%']"
+  xpath_native <- "//PIPE"
+
+  Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "file")) {
+      return(list())
+    }
+
+    xml <- source_expression$full_xml_parsed_content
+
+    match_magrittr <- xml2::xml_find_all(xml, xpath_magrittr)
+    match_native <- xml2::xml_find_all(xml, xpath_native)
+
+    n_magrittr <- length(match_magrittr)
+    n_native <- length(match_native)
+
+    if (pipe == "auto" && n_magrittr > 0L && n_native > 0L) {
+      xml_nodes_to_lints(
+        xml = c(match_magrittr, match_native),
+        source_expression = source_expression,
+        lint_message = "Use consistent pipe operators (either all %>% or all |>).",
```

Comment:
We could enhance this by customizing:

`Found %d instances of %>% and %d instances of |>. Stick to one pipe operator.`

WDYT?